### ISSUE TITLE
Accessible drag-and-drop

### ISSUE THREAD

**Issue by ndarilek** [wekan/wekan#459]
As a screen reader user, I find myself occasionally needing to reorder lists and cards. Looking at the code, I think drag-and-drop is supported but I'm not sure in which contexts this is true.

An accessible DND implementation is a bit complicated and I'm not immediately sure how to accomplish it. However, maybe some "Move up/Move down" arrows with aria-labels would be a good first implementation. They could be located in the heading for the card/board, and hidden with Bootstrap's sr-only style if you didn't want them visible (I know Bootstrap isn't being used here but you could look to its sr-only definition as an example of how to reveal something only to screen readers.)

**Comment by mfaure** [comment]
Wordpress (4.4) has a rather good accessibility implementation of drag-n-drop (in its backoffice)

**Comment by mquandalle** [comment]
I think one solution is to provide an alternative way of moving cards/tasks that doesn’t involve drag and drop. Trello has an option in their card menu to “Move a card” which provide a form where you can select a board, a list in this board, and a position in this list and press “Move”. Drag and drop cards is fun, but this more classical form would be useful for screen reader.

**Comment by ndarilek** [comment]
The issue with that is it seems to require 2 distinct steps that require 
switching between contexts. To use your system, I have to count the 
cards in the destination board, memorize the place, then input that into 
a form. If I misremember the number or had it wrong to begin with, I 
have to start the whole process over, including recounting. This seems 
somewhat more complex than simple move up/down arrows that would let me 
examine the change as I make it, directly in the relevant context.

Haven't seen WordPress 4.4's DND implementation so can't comment on it.

**Comment by mquandalle** [comment]
You don’t really need to memorize the actual position since it is the default value of the `<select>` field. Basically once your browser focus on this field you can press <kbd>↑</kbd> twice and press <kbd>Enter</kbd> to move the card up above the two previous one.

I haven’t see Wordpress 4.4 DnD neither.

**Comment by ndarilek** [comment]
Ah, it's a select. OK, that'd work.

**Comment by mquandalle** [comment]
Yes with this solution all board selection, list selection, and position selection would be implemented using `<select>` fields (or maybe enhanced select-like HTML components, in which case we need to be careful about accessibility support).

**Comment by Geniks** [comment]
any news about that ?

**Comment by mgifford** [comment]
Just wanting to check in to see if there have been any updates since 2017 on this. None of the commercial tools out there have done a good job with accessibility. I think GitHub will be pretty good for this feature.

**Comment by xet7** [comment]
@mgifford 

> Just wanting to check in to see if there have been any updates since 2017 on this. None of the commercial tools out there have done a good job with accessibility.

- I'm currently trying to learn about WAI-ARIA etc, what it would require for Wekan to be accessible. If you have more info, please add comments to this issue. Thanks!
- At [Roadmap is mentioned about already received funding for Accessibility etc](https://boards.wekan.team/b/D2SzJKZDS4Z48yeQH/wekan-open-source-kanban-board-with-mit-license/mXXFpsWPcBbNQtZEj)
- I have [small prototype with login page](https://github.com/KanRule/php/blob/master/public/index.php) with HTML4/CSS only (no Javascript), but I need to implement other pages with. To support also old browsers that do not support Javascript well, like Amiga AWeb, NetSurf, Lynx, etc, for those Wekan can not have 6 MB of Javascript at frontend. I would presume that for accessibility, mostly good enough HTML/CSS is required. Although current backend is PHP in that prototype, it could change to Javascript or something else.
- Some of my friends would also like frontend without Javascript. [Recent HN article](https://news.ycombinator.com/item?id=27484050)
- My old plans here https://github.com/KanRule/hx/wiki/Frontend-Browser-No-Javascript
- https://github.com/alphagov/accessible-autocomplete
- https://www.digitala11y.com/accessible-ui-component-libraries-roundup/
- https://github.com/salesforce-ux/dnd-a11y-patterns
- https://web.dev

> I think GitHub will be pretty good for this feature.

I don't know, is GitHub accessible?

**Comment by xet7** [comment]
@mgifford

Also there is #2499 about Dragon speech recognition sofware. Do you know does someone still use that software?

Or what do people use for screen readers, etc accessibility?

I'm still complete novice on accessibility. I'm trying to learn.

**Comment by xet7** [comment]
@mgifford 

For current Wekan, it's done with Meteor web framework, both frontend and backend are done with Javascript. But it's not necessary to run all that Javascript at frontend browserside, I most likely change most Javascript to run serverside, and have frontend with optional Javascript only.

Current Meteor Wekan supported webbrowsers are here:
https://github.com/wekan/wekan/wiki/Browser-compatibility-matrix

There Javascript drag-drop, animations, etc currently in Wekan, but those could be done some other way when there is no Javascript at frontend.

**Comment by mgifford** [comment]
Lots of stuff here.  GitHub has some accessibility features. Their kanban board has some big issues. It definitely doesn't meet WCAG compliance requirements. 

Very cool about the grant!  That's terrific. 

This isn't a problem with JS. Having a fallback is nice, but that is kinda secondary issue. Most assistive technology works just fine working with the DOM.  There are many reasons to build an HTML/CSS driven presentation, but I don't see accessibility as a big one for this. 

GatsbyJS is also pretty solidly JS, and they might be able to find some material that you could draw on here:
    https://www.gatsbyjs.com/accessibility-statement

I don't know many folks who use speech to text tools like Dragon. Apple now comes with [Voice Control ](https://www.imore.com/everything-you-can-do-voice-control-iphone-and-ipad)built in to their OSs.  This is something that's going to be an issue for people with mobility challenges, but also folks who aren't able to type but want to interact with a device.  Say you're using VR.  

@xet7 start with https://accessibilityinsights.io/ eventually start looking at building with axe into your CI like GatsbyJS does.  If you use Twitter, follow some of these folks https://github.com/joe-watkins/top-people-to-follow-in-web-accessibility

There is just so much available on GitHub projects too.  

Meteor doesn't seem to have done a great job building in accessibility. React is making this a priority though https://reactjs.org/docs/accessibility.html

Hope this helps. We're also building out some resources here https://accessibility.civicactions.com/

**Comment by xet7** [comment]
@mgifford 

I do have remote access to Mac Mini M1, thanks to MacStadium: https://github.com/wekan/wekan/wiki/Mac

How should interacting in VR work, for it to be accessible?

**Comment by xet7** [comment]
Related #337

**Comment by xet7** [comment]
@mgifford

> Meteor doesn't seem to have done a great job building in accessibility. React is making this a priority though https://reactjs.org/docs/accessibility.html

For frontend UI, Wekan currently uses Blaze. There has been new releases of Blaze, but I have not yet checked about Blaze accessibility features.

Meteor works also with other frontend frameworks. When creating new Meteor project, there are these alternatives:

```
$ meteor create
Usage: meteor create [--release <release>] [--bare|--minimal|--full|--react|--vue|--apollo|--svelte|--blaze] <path>
       meteor create [--release <release>] --example <example_name> [<path>]
       meteor create --list
       meteor create --package [<package_name>]

Options:
  --package     Create a new meteor package instead of an app.
  --example     Example template to use.
  --list        Show list of available examples.
  --bare        Create an empty app.
  --minimal     Create an app with as few Meteor packages as possible.
  --full        Create a fully scaffolded app.
  --react       Create a basic react-based app, same as default.
  --vue         Create a basic vue-based app.
  --apollo      Create a basic apollo-based app.
  --svelte      Create a basic svelte-based app.
  --typescript  Create a basic Typescript React-based app.
  --blaze       Create a basic blaze-based app.
```

**Comment by xet7** [comment]
@mgifford 

Looking at the links you provided, it's kind of too much to read. Do you have some examples what should be changed in Wekan, to make it accessible? And which ones of those recommendations and changes are necessary, for Wekan to be accessible enough? Is there someone that could test Wekan for accessibility? Like some actual user, that would be testing does something improve or not, and provide feedback?

**Comment by mgifford** [comment]
@xet7 accessibility is like security. You can make wekan more secure. You can make sure you are doing your basic due diligence to ensure that it is hard, or even very hard to exploit. Still, it will never be secure. How secure it will be totally depends on how vigilant the community of developers & users is. What culture exists around the software with regard to security. 

Same thing with accessibility. 

So don't aim to make it "just good enough". I suspect there's a long way before it could meet minimum EU government regulations for accessibility of public sites.  Aim instead to make it a bit more accessible than it was previously. 

Start with this tool https://accessibilityinsights.io as a simple browser extension.

Run it against your demo. Fix what you can.

**Comment by xet7** [comment]
@mgifford 

Thanks! I have looked at various accessibility info, and I'm slowly beginning to understand what accessibility means. Yes, it will take time.

**Comment by xet7** [comment]
More accessibility links I found:
- Newest accessibility components http://whatsock.com and https://github.com/whatsock/apex but drag drop example uses [dragula](https://github.com/bevacqua/dragula) that [was not powerful enough](https://github.com/wekan/wekan/blob/master/client/components/lists/list.js#L13-L19) because Wekan original creator mquandalle instead selected [jquery sortable](https://jqueryui.com/sortable/) that is currently used in Wekan
- Accessible JQuery UI https://hanshillen.github.io/jqtest/
- Accessible JQuery UI timepicker https://www.jqueryscript.net/time-clock/Small-Accessible-jQuery-Dropdown-Time-Picker-qcTimepicker.html
- Accessible timepicker https://trentrichardson.com/add-ui-slider-accessibility-to-timepicker.html and https://trentrichardson.com/examples/timepicker/ , related  #3873

**Comment by mgifford** [comment]
This is a bit old, but might be some useful ideas here too https://www.drupal.org/node/2920006

**Comment by xet7** [comment]
Some related article https://adrianroselli.com/2021/06/using-css-to-enforce-accessibility.html

**Comment by xet7** [comment]
It seems that accessibility surveys at https://wekan.github.io did not yet have any useful responses.

But anyway, from some other community I got this real info how to implement accessibility.

## Terminal text readable/speakable output: DOS/Windows/Linux

- There are screen readers for most operating systems
- There is JAWS for DOS at 11-minute mark of https://www.youtube.com/watch?v=bPW2S4fZMJY but JAWS is very slow on DosBox. There are some wishes to have Open Source alternative that would work in FreeDOS.
- Screen readers automatically speak any text that gets output to the terminal. So it's OK to show some amount of text, and then menu options after that, what key to press next.
- ASCII art is a problem: it's text, so screen readers speak it as it appears. So for example, we all know what `:)` means, but screen readers see it as "colon right paren", not "smiley". So any icons or smileys should be converted to readable text. As long as ASCII art is an optional feature that can be turned off, a11y is a non-issue for terminal apps.

TODO:

- [ ] In some WeKan CLI terminal based prototype try to make text-based kanban working.  Also PRs welcome.

## Web: HTML only for screen readers

- related WCAG and ARIA specs
- screen readers can see HTML but not CSS, the importance of keyboard support and so on.

Here's a quick-and-easy way to test if your website or web app is accessible:

1. Turn on your screen reader (every OS has one in settings).
2. Power off your monitor.
3. Unplug your mouse.
4. Try to use your app.
5. When you're done, run the browser extension "Axe", a great open-source tool by a company called Deque https://www.deque.com/axe/

TODO:

- [ ] Try to add some fixes to WeKan that browser extension "Axe" recommends. Also PRs welcome.
- [ ] Try making WeKan usable in Links/Lynx/ELinks etc text based browsers, with first some text that is speakable, then some menu options after that, possibility to use only keyboard etc.

**Comment by xet7** [comment]
Related:

https://scribe.rip/@r.d.t.prater/linux-accessibility-an-unmaintained-mess-8fbf9decaf8a

https://news.ycombinator.com/item?id=31375102

**Comment by xet7** [comment]
Reading this about accessibility:
https://dev.to/abbeyperini/web-development-accessibility-f8i

Above article has link to this article, that mentions that Eric Meyer’s CSS Reset should not be used:
https://dev.to/colabottles/stop-removing-focus-2o7b

WeKan includes that Eric Meyers's CSS Reset here. When converting Stylus to CSS, those extra bullet points were because that CSS Reset was missing, so it was added back.
https://github.com/wekan/wekan/blob/master/client/components/main/layouts.css#L1-L57

Related to accessibility, and adding support for more browsers, when I tested at BrowserStack (that donated access to testing with various browsers to WeKan), Multiverse WeKan static test page example at https://wekan.team/test/ , screen readers at Windows and Mac did also correctly read emojis, and load page fast. That test page loaded fast also at Mojave #4529 . Test page did load and show correctly in all browsers, only problem was if browser did not support newest TLS. So I presume future of WeKan is to be more like Multiverse WeKan so that there is support for a lot of more browsers.

When at BrowserStack I tested Meteor WeKan Roadmap page loaded very slowly, many times it did not load at all, or showed Safari being unresponsive on some older macOS, etc.

**Comment by xet7** [comment]
About many kinds of accessibility devices:
https://ericwbailey.design/writing/truths-about-digital-accessibility/

**Comment by xet7** [comment]
About accessible routing, and what was found in various accessibility testing:
https://www.gatsbyjs.com/blog/2019-07-11-user-testing-accessible-client-routing/

**Comment by xet7** [comment]
This article has some links to accessibility:
https://news.ycombinator.com/item?id=31785556

**Comment by xet7** [comment]
It seems animations should be avoided for accessibility:

https://www.theverge.com/23191768/animation-accessibility-neurodivergence

https://news.ycombinator.com/item?id=32081528

https://www.theverge.com/23188678/accessibility-disability-tech

**Comment by mgifford** [comment]
Yup..  Can make people visually sick. Visually Induced Motion Sickness (VIMS).

**Comment by xet7** [comment]
This also about accessibility for blind:

https://news.ycombinator.com/item?id=32252732

**Comment by xet7** [comment]
Smart glasses turn speech into subtitles, in real-time

https://xrai.glass

https://news.ycombinator.com/item?id=32320003

**Comment by xet7** [comment]
About low end devices:

https://shkspr.mobi/blog/2021/01/the-unreasonable-effectiveness-of-simple-html/

**Comment by xet7** [comment]
The impact of removing jQuery on web performance, and lightweight goverment website design systems:

https://insidegovuk.blog.gov.uk/2022/08/15/the-impact-of-removing-jquery-on-our-web-performance/

From https://news.ycombinator.com/item?id=32480755 some comments:

ctrlmeta 44 minutes ago:

Has anyone noticed how all UK government websites like https://www.gov.uk/, https://www.nhs.uk/, https://tfl.gov.uk/, https://coronavirus.data.gov.uk/ have the same look-and-feel and the same UX? Are there more such  examples of countries that have uniform UX for their government websites?

magnio 38 minutes ago:

Yes, this is an example of government design systems, which are becoming popular among developed countries:

- The UK: https://design-system.service.gov.uk/

- The US: https://designsystem.digital.gov/

- Canada: https://www.canada.ca/en/government/about/design-system.html

- Argentina: https://argob.github.io/poncho/

- Italy: https://designers.italia.it/

- Singapore: https://www.designsystem.tech.gov.sg/

- Estonia: https://brand.estonia.ee/

**Comment by xet7** [comment]
Hmm, I don't understand what is the point here?

- Aria-label is a code smell https://ericwbailey.website/published/aria-label-is-a-code-smell/
- https://news.ycombinator.com/item?id=33409501

**Comment by xet7** [comment]
https://support.google.com/accessibility/android/answer/6376570?hl=en

**Comment by xet7** [comment]
![accessibility1](https://user-images.githubusercontent.com/15545/210395735-19b0ebf9-4d3c-4042-803f-82e777b482b1.png)

**Comment by xet7** [comment]
![accessibility2](https://user-images.githubusercontent.com/15545/210395775-158f0222-8437-4ce1-88e3-c5abce8a5800.png)

**Comment by xet7** [comment]
- About scrollbars: https://news.ycombinator.com/item?id=37864867
- About text size etc: https://news.ycombinator.com/item?id=37902412

**Comment by xet7** [comment]
About Mobile-First/Mobile-Only:
- https://www.nngroup.com/articles/content-dispersion/
- https://news.ycombinator.com/item?id=38025767

But I do not know how this would affect WeKan, what improvements would be good? WeKan has mobile mode and desktop mode, and toggle drag handles.

Sometime maybe would be useful to add toggle mobile/desktop mode at mobile.

**Comment by xet7** [comment]
From [this article](https://dev.to/refine/htmx-exploring-the-capabilities-3gma#comment-2b2ai)
 about HTMX:

> Why does every article about htmx uses non-semantic HTML and has so much accessibility issues? I mean, most CSR app examples would use buttons here at a minimum. They wouldn't use forms most if the time but that less important because with broken/absent JS the app is just not there (blank page), but with htmx you're enhancing html and that means you just have a non-blank but entirely broken page.

I try to make WeKan more accessible.

**Comment by xet7** [comment]
About accessibility:

https://news.ycombinator.com/item?id=38796582

**Comment by xet7** [comment]
About how accessibility features are used:

https://www.w3.org/WAI/people-use-web/

**Comment by xet7** [comment]
AI and accessibility:

https://www.lambdatest.com/blog/ai-and-accessibility/

**Comment by xet7** [comment]
Show HN: SpeakMyVoice – App for people with vocal or speech difficulties 

https://speakmyvoice.com

https://news.ycombinator.com/item?id=41846623

**Comment by xet7** [comment]
Bruno Prieto - Making accessible web apps with Rails and Hotwire - Rails World 2024

https://www.youtube.com/watch?v=OK2wApjyswM

Notes:
- Test with screen readers at Windows/Mac/Linux
- Use HTML links to move to different page
- Use HTML buttons to do some action
- Do not make custom components
- For hiding something, use `hide` at that form element
- Use HTML provided forms and HTTP POST to server
- Use semantic HTML, divide to sections
- Use H1 etc tags
- Check keyboard navigation order
- Check that all content is readable with screen reader
- Etc

**Comment by xet7** [comment]
Accessibility: Don't Use Fake Bold or Italic in Social Media

https://adrianroselli.com/2025/03/dont-use-fake-bold-or-italic-in-social-media.html

https://news.ycombinator.com/item?id=43302835

**Comment by xet7** [comment]
Accessibility distros, for example https://slint.fr

This info from:
- https://fireborn.mataroa.blog/blog/i-want-to-love-linux-it-doesnt-love-me-back-post-1-built-for-control-but-not-for-people/
- https://news.ycombinator.com/item?id=43945373

**Comment by xet7** [comment]
Moved to here from #5778

From @Adamsss001

Accessibility improvements include better keyboard navigation, screen reader compatibility, and high-contrast options, ensuring Wekan is usable and welcoming for everyone, regardless of ability or preferred working conditions.

**Comment by xet7** [comment]
## An accessibility update – GTK Development Blog

- https://blog.gtk.org/2025/05/12/an-accessibility-update/
- https://news.ycombinator.com/item?id=43971685

**Comment by xet7** [comment]
Q: Does Apple new Liquid Glass design affect accessibility?

A: (by Google Gemini)

Apple's new "Liquid Glass" design, introduced with iOS 26 and other operating systems, aims to create a more dynamic, translucent, and expressive user interface. While visually striking, it does have potential implications for accessibility, particularly for users with visual impairments.

Here's a breakdown of how it affects accessibility:

* **Increased Transparency and Legibility:** The core of Liquid Glass is its transparent, glassy look for icons, menus, and windows. While this can create a sense of depth and focus on content, some users, especially those with visual impairments, might find it **difficult to read text or view certain elements** when the background is busy or has low contrast. The transparency can make content clarity suffer.

* **Accessibility Features as a Solution:** Apple is aware of these potential issues and has emphasized that Liquid Glass works with existing accessibility features. Specifically, the **"Reduce Transparency" setting** (found in Settings > Accessibility > Display & Text Size on iOS/iPadOS, and System Settings > Accessibility > Display on macOS) is crucial. Toggling this on replaces the transparent effect with a solid background, significantly boosting contrast and visibility. This setting existed before Liquid Glass and was designed to help users with similar vision issues.

* **Dynamic Adaptation:** Apple states that Liquid Glass is designed to be highly adaptable. It responds automatically to user preferences, simplifying transitions under "Reduce Motion," adding opacity when "Reduce Transparency" is enabled, and increasing separation when "Increase Contrast" is on. It also dynamically adjusts shadow opacity, blur density, and brightness based on the underlying content.

* **Developer Guidelines:** Apple's developer documentation advises developers to test their interfaces with accessibility settings. It emphasizes that standard components from system frameworks will adapt automatically, but custom elements and animations need to provide a good fallback experience when these settings are on. Developers are also cautioned against overusing Liquid Glass effects to avoid distracting from content.

* **Broader Accessibility Enhancements:** Alongside the Liquid Glass design, Apple is introducing a range of new accessibility features in its upcoming OS updates. These include:
    * **Magnifier for Mac:** A new vision tool that connects to a user's camera to zoom in on surroundings.
    * **Accessibility Reader:** A system-wide reading mode to make text easier to read for users with various disabilities, offering extensive options for font, color, spacing, and Spoken Content.
    * **Braille Access:** An all-new experience that turns Apple devices into full-featured braille note takers.
    * **Live Captions on Apple Watch:** Real-time transcriptions for users who are deaf or hard of hearing.

In conclusion, while the aesthetic of Apple's Liquid Glass design might present challenges for some users with visual impairments due to its transparency, Apple has built in mechanisms, particularly the "Reduce Transparency" setting, to mitigate these issues. Furthermore, the company continues to introduce a broader suite of accessibility features to enhance usability for all users.

**Comment by xet7** [comment]
## Selfish reasons for building accessible UIs

- https://nolanlawson.com/2025/06/16/selfish-reasons-for-building-accessible-uis/
- https://news.ycombinator.com/item?id=44294853

**Comment by xet7** [comment]
## Don't put aria-label on generic elements like divs

- https://www.matuzo.at/blog/2026/aria-label-generic-elements
- https://news.ycombinator.com/item?id=48237159

**Comment by xet7** [comment]
Fixed in the upcoming WeKan release. See CHANGELOG.md Upcoming section.
